File: kangaroo_app_sdk_android/android/src/main/kotlin/com/kangaroo/flutterplugin/android/features/getSpinToWinPrizes/GetSpinToWinPrizesHandler.kt
@file:Suppress("INLINE_FROM_HIGHER_PLATFORM")
@file:OptIn(ExperimentalJsExport::class)
package com.kangaroo.flutterplugin.android.features.getSpinToWinPrizes

import com.kangaroo.flutterplugin.android.base.PluginChannelHandler
import com.kangaroo.flutterplugin.android.base.pushSerializedResultToEventSink
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import kotlin.js.ExperimentalJsExport
import features.getSpinToWinPrizes.GetSpinToWinPrizesApi
import features.getSpinToWinPrizes.models.SpinToWinPrizesModel

import features.getSpinToWinPrizes.serializeGetSpinToWinPrizesState
import kangaroorewards.appsdk.core.domain.SerializedResult
import kangaroorewards.appsdk.core.domain.toJsonResult


class GetSpinToWinPrizesHandler : EventChannel.StreamHandler, PluginChannelHandler {
    var sink: EventChannel.EventSink? = null

    override val methodChannel: String
        get() = "customer_sdk/methods/get_spin_to_win_prizes"

    override val eventChannel: String
        get() = "customer_sdk/events/get_spin_to_win_prizes"

    override suspend fun onMethodCall(call: MethodCall): String? {
        return getSpinToWinPrizes(call)
    }

    override fun getStreamHandler(): EventChannel.StreamHandler {
        return this
    }

    companion object {
        suspend fun getSpinToWinPrizes(call: MethodCall): String? {
            val result = GetSpinToWinPrizesApi().getSpinToWinPrizes(
                businessId = call.argument<String>("businessId") as String
            ).toJsonResult<SpinToWinPrizesModel>()

            return when (result) {
                is SerializedResult.Success -> result.data
                is SerializedResult.UnauthorizedError -> result.error
                is SerializedResult.UnknownError -> result.error
                else -> null
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        GetSpinToWinPrizesApi().getSpinToWinPrizesState.serializeGetSpinToWinPrizesState().observe { result ->
            pushSerializedResultToEventSink<GetSpinToWinPrizesApi>(sink, result)
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}

File: kangaroo_app_sdk_ios/ios/Classes/features/getSpinToWinPrizes/GetSpinToWinPrizesHandler.swift
import Foundation
import Flutter
import KangarooAppSdkCustomer

class GetSpinToWinPrizesHandler: NSObject, FlutterStreamHandler, PluginChannelHandler {
    var sink: FlutterEventSink?

    var methodChannel: String = "customer_sdk/methods/get_spin_to_win_prizes"

    var eventChannel: String = "customer_sdk/events/get_spin_to_win_prizes"

    func onMethodCall(call: FlutterMethodCall) async -> Any? {
        return await GetSpinToWinPrizesHandler.getSpinToWinPrizes(call: call)
    }

    func getStreamHandler() -> (FlutterStreamHandler & NSObjectProtocol)? {
        return self
    }


    static func getSpinToWinPrizes(call: FlutterMethodCall) async -> String? {
        

        


        

        guard let args = call.arguments else {
            return nil
        }
        do {
       if let myArgs = args as? [String: Any] {
                        guard let businessId = myArgs["businessId"] as? String else {return nil}

         let result = try await GetSpinToWinPrizesApi().getSpinToWinPrizes(
                businessId: businessId
            ).serializeGetSpinToWinPrizesApiResult()

        switch result {
            case let result as SerializedResultSuccess:
                return result.data
            case let result as SerializedResultUnauthorizedError:
                return result.error
            case let result as SerializedResultUnknownError:
                return result.error
            default:
                return nil
                }
            }
        }
        catch {
            return nil
        }
        
        return nil
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping
        FlutterEventSink) -> FlutterError? {
        sink = events
        GetSpinToWinPrizesApi().getSpinToWinPrizesState
            .serializeGetSpinToWinPrizesState().observe { result in
            switch result {
            case let result as SerializedResultIdle:
                self.sink?(result.state)
            case let result as SerializedResultLoading:
                self.sink?(result.state)
            case let result as SerializedResultSuccess:
                self.sink?(result.data)
            case let result as SerializedResultEmptyResponse:
                self.sink?(result.body)
            case let result as SerializedResultUnauthorizedError:
                self.sink?(result.error)
            case let result as SerializedResultUnknownError:
                self.sink?(result.error)
            default:
                break
            }
        }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}
